src/osd/OSDMap.cc

```diff
@@ -4541,13 +4543,34 @@ int OSDMap::calc_pg_upmaps(
   int total_pgs = 0;
   float osd_weight_total = 0;
   map<int,float> osd_weight;
+  int64_t pg_num = 0;
+  for (auto& i : pools) {
+    if (!only_pools.empty() && !only_pools.count(i.first))
+      continue;
+    pg_num += i.second.get_pg_num();
+  }
+  auto pgs_per_chunk = cct->_conf.get_val<int64_t>(
+    "mon_osd_mapping_pgs_per_chunk");
+  OSDMapMapping mapping; // pg <-> osd mappings
+  bool prime_pg_mapping = false;
+  // parallel CRUSH calculation if necessary
+  if (pg_num > pgs_per_chunk && tp) {
+    prime_pg_mapping = true;
+    ParallelPGMapper mapper(cct, tp);
+    auto mapping_job = mapping.start_update(tmp, mapper, pgs_per_chunk,
+      only_pools.empty() ? nullptr : &only_pools);
+    mapping_job->wait();
```

Comment:
I'm not quite following the concept here.  We push the work to another thread, but this thread still blocks.. how is that different from doing the work in this thread?

Reply:
> We push the work to another thread

not another thread, but a thread pool that may contain multiple threads. So the calculation could be done at a faster speed.